### PR TITLE
Support generic function in `generate_function` assist

### PR DESCRIPTION
**Pull request by lowr**
Part of #3639

This PR adds support for generic function generation in `generate_function` assist. Now the assist looks for generic parameters and trait bounds in scope, filters out irrelevant ones, and generates new function with them.

See `fn_generic_params()` for the outline of the procedure, and see comments on `filter_unnecessary_bounds()` for criteria for filtering. I think it's good criteria for most cases, but I'm open to opinions and suggestions.

The diff is pretty big, but it should run in linear time w.r.t. the number of nodes we operate on and should be fast enough.

Some notes:
- When we generate function in an existing impl, generic parameters may cause name conflict. While we can detect the conflict and rename conflicting params, I didn't find it worthwhile mainly because it's really easy to resolve on IDE: use Rename functionality.
- I've implemented graph structure myself, because we don't have graph library as a dependency and we only need the simplest one.
  - Although `petgraph` is in our dependency graph and I was initially looking to use it, we don't actually depend on it AFAICT since it's only used in chalk's specialization graph handling, which we don't use. I'd be happy to replace my implementation with `petgraph` if it's okay to use it though.
- There are some caveats that I consider out of scope of this PR. See FIXME notes on added tests.

### CODE REVIEW

File: crates/hir-ty/src/lib.rs [hunk 1]
```diff
@@ -562,3 +564,68 @@ pub fn callable_sig_from_fnonce(
...
+    ) -> std::ops::ControlFlow<Self::BreakTy> {
+        if let chalk_ir::ConstValue::Placeholder(idx) = constant.data(Interner).value {
+            self.collect(idx);
+        }
+        std::ops::ControlFlow::Continue(())
+    }
+}
+
+/// Returns unique placeholders for types and consts contained in `value`.
+pub fn collect_placeholders<T>(value: &T, db: &dyn HirDatabase) -> Vec<TypeOrConstParamId>
```
Comment (lowr) [inline crates/hir-ty/src/lib.rs:624]: I'm not sure where to place this function. `hir_ty::utils` says it's for "Helper functions for working with def, which (...) can't be computed directly from `*Data`" and it doesn't seem to fit there?

File: crates/ide-assists/src/handlers/generate_function.rs [hunk 2]
```diff
@@ -473,37 +515,385 @@ impl GeneratedFunctionTarget {
...
+
+    let generic_params: Vec<_> =
+        generic_params.into_iter().map(|it| it.node.clone_for_update()).collect();
+    let where_preds: Vec<_> =
+        where_preds.into_iter().map(|it| it.node.clone_for_update()).collect();
+
+    // 4. Rewrite paths
+    if let Some(param) = generic_params.first() {
+        let source_scope = ctx.sema.scope(param.syntax()).unwrap();
+        let target_scope = ctx.sema.scope(&target.parent()).unwrap();
```
Comment (lowr) [inline crates/ide-assists/src/handlers/generate_function.rs:596]: I think these are safe to `unwrap()` at this point, but let me know if it's not. When does `scope()` return `None`?

Comment (Veykril) [reply crates/ide-assists/src/handlers/generate_function.rs:596]: Scope returns `None` in rust files that are not part of the crate graph, its probably better to exit out explicitly here.

**Comment by Veykril** [comment]
r=me with the 2 unwraps replaced
@bors delegate+

**Comment by lowr** [comment]
Two rather random thoughts:
- `extract_function` assist may benefit from functions this PR adds
- I feel it might make sense to add a variant of the assist: "Generate function/method without trait bounds"

@bors r=Veykril

**Pull request merged**